src/lib/threadpool.c

```diff
@@ -553,6 +556,15 @@ int pool_init(pool_t *pool,
 	}
 
 	pool_threads_init(pool);
+
+	rc = uv_key_create(&thread_identifier_key);
```

Comment:
I think this needs to be wrapped in `uv_once`; the man page for `pthread_key_create` says it's the programmer's responsibilty to not call it more than once on a given address, and that can happen with this code if you have multiple nodes running in a process, or even if you create/run multiple nodes serially.

Reply:
You are right! I don't know why but my local man pages didn't have information about that but when I look at https://linux.die.net/man/3/pthread_key_create there is a whole section about it

Reply:
I don't know about multiple threads but creation of multiple nodes (hence pools) within a process happens in the test suite. Maybe it's harmless to call `uv_key_create` more than once on the same key with the same value as long as there's no race, but the language of the man page seems stricter than that:

Reply:
I do think that ideally it should be safe to call `dqlite_node_create` from multiple threads without synchronization.

Reply:
I also don't know enough about tls to know if it is safe to create the same key more than once. I agree that ideally we should be able to call `dqlite_node_create` from multiple threads without doing anything special. But then we have to change the approach because there is no single "main thread". What we can do instead is to record whether a thread belongs to the threadpool or not, which is conceptually easier when dealing with multiple pools.

I don't think we need to record the provenance in regards to which pool spawned which thread, but what do you think?

Reply:
Take a look at https://github.com/canonical/dqlite/pull/673/commits/6d2669609db19c249dfb0af27d2e2295be1f4e75, I have changed the code to use `uv_once` and also to identify the pool threads.

Reply:
Yeah, that is unfortunate.

Reply:
I thought about that as well but that would be the complement of the threads that are part of the pool, right? You can always negate the assertion.

Reply:
Well, you could set one magic number for loop threads and another for pool threads, right?

Reply:
Discussed offline, we are going to postpone discussing this until a need arrives to identify specific threads, e.g. ones running a specific loop.